File: taskmanagement-ui/src/features/users/api/RoleService.ts
import { apiClient } from '@/api/axios';
import type { ApiResponse } from '@/features/auth/types/auth';
import type { PageResponse } from '@/features/users/types/user';

export interface Role {
  id: number;
  name: string;
  description?: string;
}

export const RoleService = {
  getRoles: async (page: number, size: number): Promise<PageResponse<Role>> => {
    const response = await apiClient.get<ApiResponse<PageResponse<Role>>>('/api/roles/paging', {
      params: { page, size }
    });
    return response.data.data;
  },

  searchRoles: async (keyword: string, page: number, size: number): Promise<PageResponse<Role>> => {
    const response = await apiClient.get<ApiResponse<PageResponse<Role>>>('/api/roles/search', {
      params: { keyword, page, size }
    });
    return response.data.data;
  },

  getAllRoles: async (): Promise<Role[]> => {
    const response = await apiClient.get<ApiResponse<Role[]>>('/api/roles');
    return response.data.data;
  },

  createRole: async (name: string): Promise<Role> => {
    const response = await apiClient.post<ApiResponse<Role>>('/api/roles', { name });
    return response.data.data;
  },

  updateRole: async (id: number, name: string): Promise<Role> => {
    const response = await apiClient.put<ApiResponse<Role>>(`/api/roles/${id}`, { name });
    return response.data.data;
  },

  deleteRole: async (id: number): Promise<void> => {
    await apiClient.delete(`/api/roles/${id}`);
  },

  assignRoleToUser: async (userId: number, roleId: number): Promise<void> => {
    await apiClient.post(`/api/users/${userId}/roles/${roleId}`);
  }
};

File: src/main/java/com/phong/taskmanagement/service/interfaces/RoleService.java
package com.phong.taskmanagement.service.interfaces;

import java.util.List;

import org.springframework.data.domain.Page;

import com.phong.taskmanagement.dto.request.CreateRoleRequest;
import com.phong.taskmanagement.domain.entity.Role;

public interface RoleService {

    Role createRole(CreateRoleRequest request);

    List<Role> getAllRoles();

    Role getRoleById(Long id);

    void deleteRole(Long id);

    Role updateRole(
            Long id,
            CreateRoleRequest request
    );

    Page<Role> searchRoles(
            String keyword,
            int page,
            int size
    );

    Page<Role> getRolesWithPaging(
            int page,
            int size
    );
}

